proxy/http/parser/tls.py

```diff
@@ -0,0 +1,431 @@
+                    # parse server hello done
+                    self.server_hello_done = TlsServerHelloDone()
+                    self.server_hello_done.parse(payload)
+                elif self.msg_type == tlsHandshakeType.CERTIFICATE_VERIFY:
+                    # parse certificate verify
+                    self.certificate_verify = TlsCertificateVerify()
+                    rself.certificate_verify.parse(payload)
+                elif self.msg_type == tlsHandshakeType.CLIENT_KEY_EXCHANGE:
+                    # parse client key exchange
+                    self.client_key_exchange = TlsClientKeyExchange()
+                    self.client_key_exchange.parse(payload)
+                elif self.msg_type == tlsHandshakeType.FINISHED:
+                    # parse finished
+                    self.finished = TlsFinished()
+                    self.finished.parse(payload)
+                return True, raw[4 + payload_length:]
+
+    def build(self) -> bytes:
+        data = bytes()
+        data += bytes([self.msg_type])
+        payload = bytes()
+        if self.msg_type == tlsHandshakeType.CLIENT_HELLO:
+            payload = self.client_hello.build()
+        elif self.msg_type == tlsHandshakeType.SERVER_HELLO:
+            payload = self.server_hello.build()
+        elif self.msg_type == tlsHandshakeType.CERTIFICATE:
+            payload = self.certificate.build()
+        elif self.msg_type == tlsHandshakeType.SERVER_KEY_EXCHANGE:
+            payload = self.server_key_exchange.build()
+        # calculate length
+        length = struct.pack('!I', len(payload))[1:]
+        data += length
+        data += payload
+        return data
+
+
+class TlsParser:
+    """TLS packet parser"""
+
+    def __init__(self) -> None:
+        self.content_type: int = tlsContentType.OTHER
+        self.protocol_version: Optional[TlsProtocolVersion] = None
+        self.length: Optional[bytes] = None
+        # only parse hand shake payload temporary
+        self.handshake: Optional[TlsHandshake] = None
+        self.certificate: Optional[TlsCertificate]
+
+    def parse(self, raw: bytes) -> (bool, bytes):
+        """parse TLS fragmentation
+        Ref: https://datatracker.ietf.org/doc/html/rfc5246#page-15
+             https://datatracker.ietf.org/doc/html/rfc5077#page-3
+             https://datatracker.ietf.org/doc/html/rfc8446#page-10
+        """
+        length = len(raw)
+        if length < 5:
+            logger.debug('invalid data, len(raw) = %s', length)
+            return False, raw
+        else:
+            payload_length, = struct.unpack('!H', raw[3:5])
+            self.protocol_version
+            if length < 5 + payload_length:
+                logger.debug('incomplete data, len(raw) = %s, len(payload) = %s', length, payload_length)
+                return False, raw
+            else:
+                # parse
+                self.content_type = raw[0]
+                self.protocol_version = raw[1:3]
```

Comment:
Here `self.protocol_version: Optional[TlsProtocolVersion] = None` is a `TlsProtocolVersion` type but you are assigning a byte to it here.

Reply:
Sorry, TlsParser is completed yet, we can declare self.protocol_version as bytes